cdap-formats/src/main/java/co/cask/cdap/format/DelimitedStringsRecordFormat.java

```
@@ -114,15 +134,118 @@ private boolean isStringArray(Schema schema) {
+      try {
+        mapping.put(fieldName, Integer.parseInt(fieldIndexString));
+      } catch (NumberFormatException e) {
+        throw new IllegalArgumentException("mapping keys must be integer indices");
+      }
+    }
+
+    return mapping;
+  }
+
+  /**
+   * Makes a {@link StructuredRecord} in {@link DelimitedStringsRecordFormat#read(StreamEvent)}.
+   */
+  private interface RecordMaker {
+    StructuredRecord make(Schema schema, Iterator<String> bodyFields);
+  }
+
+  /**
+   * {@link RecordMaker} that uses the default schema.
+   */
+  private static class DefaultRecordMaker implements RecordMaker {
+
+    @Override
+    public StructuredRecord make(Schema schema, Iterator<String> bodyFields) {
+      StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+
+      List<String> fields = Lists.newArrayList(bodyFields);
+      builder.set("body", fields.toArray(new String[fields.size()]));
+
+      return builder.build();
+    }
+  }
+
+  /**
+   * {@link RecordMaker} that uses a schema.
+   */
+  private static class SchemaRecordMaker implements RecordMaker {
+
+    @Override
+    public StructuredRecord make(Schema schema, Iterator<String> bodyFields) {
+      StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+      Iterator<Schema.Field> fieldsIterator = schema.getFields().iterator();
+      while (fieldsIterator.hasNext()) {
+        Schema.Field field = fieldsIterator.next();
+        Schema fieldSchema = field.getSchema();
+        String fieldName = field.getName();
+        if (isStringArray(fieldSchema)) {
+          if (!fieldsIterator.hasNext()) {
+            // only do varargs-style string array parsing on bodyField if it's the last field
+            List<String> fields = Lists.newArrayList(bodyFields);
+            builder.set(fieldName, fields.toArray(new String[fields.size()]));
+          } else {
+            // otherwise, parse bodyField into a single element array
```

yeah this should throw UnexpectedFormatException